### PR TITLE
fix-#3: refactored-backend-with-typescript

### LINKED ISSUE #3: Refactoring backend using typescript

### PR DESCRIPTION
**Pull request by vamsidhar-914**
## Summary

refactored backend with typescript

## Description
migrated all js files to ts files with all the types.

## Issue(s) Addressed
Closes #3 

## Prerequisites

- [ yes] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by vamsidhar-914** [comment]
@krishnaacharyaa 
check it out....tell me if any mistakes to be taken care of..

for tests I might need some help.I didnt touched tests for refactoring

File: backend/types/user-Type.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import mongoose from 'mongoose';
+
+export interface IUser extends Document {
+  _id: string;
+  userName: string;
+  fullName: string;
+  email: string;
+  password: string;
+  avatar?: string;
```
Comment (krishnaacharyaa) [inline backend/types/user-Type.ts:9]: Give here password string | undefined, so that it will let you define undefined

File: backend/types/user-Type.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import mongoose from 'mongoose';
+
+export interface IUser extends Document {
+  _id: string;
```
Comment (krishnaacharyaa) [inline backend/types/user-Type.ts:4]: Use types instead of interface and remove the I prefix in the naming

File: backend/controllers/auth-controller.ts [hunk 3]
```diff
@@ -38,20 +40,27 @@ export const signUpWithEmail = asyncHandler(async (req, res) => {
...
-  const accessToken = await user.generateAccessToken();
-  const refreshToken = await user.generateRefreshToken();
+  const accessToken: string = await user.generateAccessToken();
+  const refreshToken: string = await user.generateRefreshToken();
 
   user.refreshToken = refreshToken;
 
   await user.save();
-  user.password = undefined;
+  // user.password = undefined;
```
Comment (krishnaacharyaa) [inline backend/controllers/auth-controller.ts:64]: Make the password field nullable in the type, so that you don't have to comment it for the ts error

File: backend/controllers/auth-controller.ts [hunk 4]
```diff
@@ -95,7 +104,6 @@ export const signInWithEmailOrUsername = asyncHandler(async (req, res) => {
 
   user.refreshToken = refreshToken;
   await user.save();
-  user.password = undefined;
```
Comment (krishnaacharyaa) [inline backend/controllers/auth-controller.ts:107]: Same, get it back

File: backend/controllers/auth-controller.ts [hunk 5]
```diff
@@ -164,12 +178,6 @@ export const isLoggedIn = asyncHandler(async (req, res) => {
       console.log(error);
     }
   }
-  const user = await User.findById(_id);
-  if (!user) {
```
Comment (krishnaacharyaa) [inline backend/controllers/auth-controller.ts:168]: Why have we tottally removed this? please don't remove some functionality just because it is giving error XD

File: backend/controllers/posts-controller.ts [hunk 6]
```diff
@@ -154,23 +185,30 @@ export const deletePostByIdHandler = async (req, res) => {
...
 
-export const getRelatedPostsByCategories = async (req, res) => {
+export const getRelatedPostsByCategories = async (req: Request, res: Response) => {
   const { categories } = req.query;
   if (!categories) {
-    return res
-      .status(HTTP_STATUS.NOT_FOUND)
-      .json({ message: RESPONSE_MESSAGES.POSTS.CATEGORIES_NOTFOUND });
+    return res.status(HTTP_STATUS.NOT_FOUND).json({ message: 'Categories not Found' });
   }
```
Comment (krishnaacharyaa) [inline backend/controllers/posts-controller.ts:200]: Please don't hard-code, or change any variables, keep it existing as whatever it was, just do the ts refactor without any changes to the actual functionality of the code

File: backend/middlewares/auth-middleware.ts [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
+  next: NextFunction
+) => {
   const token = req.cookies?.access_token;
   if (!token) {
     return next(new ApiError(HTTP_STATUS.BAD_REQUEST, RESPONSE_MESSAGES.USERS.RE_LOGIN));
   }
 
   if (token) {
-    await jwt.verify(token, JWT_SECRET, (error, payload) => {
+    jwt.verify(token, JWT_SECRET!, (error: any, payload: any) => {
```
Comment (krishnaacharyaa) [inline backend/middlewares/auth-middleware.ts:19]: Why did you remove await

File: backend/middlewares/error-middleware.ts [hunk 8]
```diff
@@ -1,6 +1,19 @@
-import { HTTP_STATUS, RESPONSE_MESSAGES } from '../utils/constants.js';
+import { NextFunction, Request, Response } from 'express';
+import { HTTP_STATUS, RESPONSE_MESSAGES } from '../utils/constants';
 
-const errorMiddleware = (err, req, res, next) => {
+export type ErrorMiddlewareType = {
+  stack: string;
```
Comment (krishnaacharyaa) [inline backend/middlewares/error-middleware.ts:5]: If we are using `ErrorMiddlewareType` anywhere other than this file, then we have to refactor this to go and sit in the types folder, if not remove the export keyword from the type

File: backend/models/post.ts [hunk 9]
```diff
@@ -1,7 +1,8 @@
-import { Schema, model } from 'mongoose';
+import { Model, Schema, model } from 'mongoose';
+import { IPost } from '../types/post-Type';
 
 const postSchema = new Schema({
-  authorName: String,
+  authorfName: String,
   title: String,
```
Comment (krishnaacharyaa) [inline backend/models/post.ts:6]: Why fName? keep it same please

File: backend/package.json [hunk 10]
```diff
@@ -1,5 +1,5 @@
 {
-  "type": "module",
+  "type": "commonjs",
```
Comment (krishnaacharyaa) [inline backend/package.json:2]: Why? this change

File: backend/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{
+  "compilerOptions": {
+    "target": "ES2016",
+    "module": "CommonJS",
+    "outDir": "./dist",
```
Comment (krishnaacharyaa) [inline backend/tsconfig.json:5]: commonJs or module? figure this out, we don't want to go backwards we want to use "module" in the nodejs app but try to figure out what option to give in the tsconfig
https://stackoverflow.com/questions/55471795/what-is-module-option-in-tsconfig-used-for

File: backend/types/api-Error-Type.ts
Comment (krishnaacharyaa) [inline backend/types/api-Error-Type.ts:1]: Name of the file is violating guidelines, please do changes everywhere

It shouldbe api-error-type

File: backend/types/express/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { IUser } from '../models/user';
+
+declare global {
+  namespace Express {
+    interface Request {
+      user?: {
```
Comment (krishnaacharyaa) [inline backend/types/express/index.d.ts:6]: Is it truly required ? Because here we assume the default Request type if of the user type right? I prefer using User type instead of Request type , because request can be multiple we can gloablize it to user

File: backend/types/post-Type.ts
Comment (krishnaacharyaa) [inline backend/types/post-Type.ts:1]: name of the file post-type

File: backend/types/post-Type.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import { IUser } from './user-Type';
+
+export interface IPost extends Document {
+  _id?: string;
```
Comment (krishnaacharyaa) [inline backend/types/post-Type.ts:4]: make it type and why does it extend Document? who is document and rename we don't need I prefix

File: backend/types/request-User-Type.ts
Comment (krishnaacharyaa) [inline backend/types/request-User-Type.ts:1]: name

File: backend/types/request-User-Type.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import { Request } from 'express';
+
+interface RequestUserType {
+  _id: string;
```
Comment (krishnaacharyaa) [inline backend/types/request-User-Type.ts:4]: convert to type

File: backend/types/request-User-Type.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+  role: string;
+  userName: string;
+  email: string;
+}
+
+export interface RequestWithUserRole extends Request {
+  user?: RequestUserType;
```
Comment (krishnaacharyaa) [inline backend/types/request-User-Type.ts:11]: Do we actually need this? Can't we directly use the above ? make it both to type

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
HI @vamsidhar-914 appreciate the efforts, kindly don't make any changes to the codebase, just the refactoring part to the ts and don't remove the codes or comment if you get errors, kindly address all the reviews :)

File: backend/controllers/auth-controller.ts (hunk 5 above)
Comment (vamsidhar-914) [reply backend/controllers/auth-controller.ts:168]: not remove d but declared on above

File: backend/types/request-User-Type.ts (hunk 15 above)
Comment (vamsidhar-914) [reply backend/types/request-User-Type.ts:11]: in middleware req.user , user is not found in Request type...extended the Request Type added user filed

File: backend/models/post.ts (hunk 9 above)
Comment (vamsidhar-914) [reply backend/models/post.ts:6]: sorry, i didnt know that. some errors by mistakely done,i'll fix

**Comment by vamsidhar-914** [comment]
hi @krishnaacharyaa 

1.) when i use default Request type for req. it throws me error req.user is not found.so i declared index.d.ts file extended Request type with user field. but again it throws me the same error

Screenshot (218)

so i created a new type for Request and replaced it with that type , and it works fine 
Screenshot (219)

2.) when i use the type 'module' in package.json it throws this error, 
Screenshot (215)


these are the errors

**Comment by krishnaacharyaa** [comment]
@vamsidhar-914 please google these, it will be available.
1. might be because Request is a wrong import
2. because you have to change the tsconfig.json even, we can't change it from module to commonjs, no going back.

And the request type as discussed would not be actually required, try to come up with alternative

**Comment by krishnaacharyaa** [comment]
@vamsidhar-914 Hi, Are we good or blocked? 
Let's have a call and get this in, It is a important feature we need this in, early it is, better it is :)

**Comment by vamsidhar-914** [comment]
@krishnaacharyaa  sorry! been busy lately.....everything is done except for the type module issue...searched google but couldn't find it.

**Comment by krishnaacharyaa** [comment]
No worries can we arrange a call we'll debug

**Comment by vamsidhar-914** [comment]
can we connect tomorrow morning?

**Comment by krishnaacharyaa** [comment]
Nights or sat Sun 
I have office other times

**Comment by vamsidhar-914** [comment]
sure....lets connect on sunday

**Comment by krishnaacharyaa** [comment]
@vamsidhar-914 the earlier we get this, better it is :)
Now we have conflicts, it will increase as the time progresses
